go/worker/storage/p2p/checkpointsync/client.go

```diff
@@ -0,0 +1,144 @@
+	minProtocolPeers = 5
+
+	// totalProtocolPeers is the number of peers we want to have connected for checkpoint sync protocol.
+	totalProtocolPeers = 10
+)
+
+// Client is a checkpoint sync protocol client.
+type Client interface {
+	// GetCheckpoints returns a list of checkpoint metadata for all known checkpoints.
+	GetCheckpoints(ctx context.Context, request *GetCheckpointsRequest) ([]*Checkpoint, error)
+
+	// GetCheckpointChunk requests a specific checkpoint chunk.
+	GetCheckpointChunk(
+		ctx context.Context,
+		request *GetCheckpointChunkRequest,
+		cp *Checkpoint,
+	) (*GetCheckpointChunkResponse, rpc.PeerFeedback, error)
+}
+
+// Checkpoint contains checkpoint metadata together with peer information.
+type Checkpoint struct {
+	*checkpoint.Metadata
+
+	// Peers are the feedback structures of all the peers that have advertised this checkpoint.
+	Peers []rpc.PeerFeedback
+}
+
+type client struct {
+	rc          rpc.Client
+	mgr         rpc.PeerManager
+	fallbackMgr rpc.PeerManager
+}
+
+func (c *client) GetCheckpoints(ctx context.Context, request *GetCheckpointsRequest) ([]*Checkpoint, error) {
+	var rsp GetCheckpointsResponse
+	rsps, pfs, err := c.rc.CallMulti(ctx, c.getBestPeers(), MethodGetCheckpoints, request, rsp)
+	if err != nil {
+		return nil, err
+	}
+
+	// Combine deduplicated results into a single result.
+	var checkpoints []*Checkpoint
+	cps := make(map[hash.Hash]*Checkpoint)
+	for i, peerRsp := range rsps {
+		peerCps := peerRsp.(*GetCheckpointsResponse).Checkpoints
+
+		for _, cpMeta := range peerCps {
+			h := cpMeta.EncodedHash()
+			cp := cps[h]
+			if cp == nil {
+				cp = &Checkpoint{
+					Metadata: cpMeta,
+				}
+				cps[h] = cp
+				checkpoints = append(checkpoints, cp)
+			}
+			cp.Peers = append(cp.Peers, pfs[i])
+		}
+
+		// Record success for a peer if it returned at least one checkpoint.
+		if len(peerCps) > 0 {
+			pfs[i].RecordSuccess()
+		}
+	}
+	return checkpoints, nil
+}
+
+func (c *client) GetCheckpointChunk(
+	ctx context.Context,
+	request *GetCheckpointChunkRequest,
+	cp *Checkpoint,
+) (*GetCheckpointChunkResponse, rpc.PeerFeedback, error) {
+	var opts []rpc.BestPeersOption
+	// When a checkpoint is passed, we limit requests to only those peers that actually advertised
+	// having the checkpoint in question to avoid needless requests.
+	if cp != nil {
+		peers := make([]core.PeerID, 0, len(cp.Peers))
+		for _, pf := range cp.Peers {
+			peers = append(peers, pf.PeerID())
+		}
+		opts = append(opts, rpc.WithLimitPeers(peers))
+	}
+
+	var rsp GetCheckpointChunkResponse
+	peers := c.getBestPeers(opts...)
+	pf, err := c.rc.CallOne(ctx, peers, MethodGetCheckpointChunk, request, &rsp,
+		rpc.WithMaxPeerResponseTime(MaxGetCheckpointChunkResponseTime),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &rsp, pf, nil
+}
+
+func (c *client) getBestPeers(opts ...rpc.BestPeersOption) []core.PeerID {
+	return append(c.mgr.GetBestPeers(opts...), c.fallbackMgr.GetBestPeers(opts...)...)
```

Comment:
I'm not sure if this will work as expected. Maybe we should have just 2 clients, everyone with its own protocol ID and then we can switch to the legacy one in methods like `GetCheckpointChunk` if we don't get an answer from the first one.